src/sentry/api/authentication.py

```diff
@@ -69,6 +70,40 @@ def authenticate_credentials(self, userid, password):
         return (AnonymousUser(), key)
 
 
+class ClientIdSecretAuthentication(QuietBasicAuthentication):
+    """
+    Authenticates a Sentry Application using its Client ID and Secret
+
+    This will be the method by which we identify which Sentry Application is
+    making the request, for any requests not scoped to an installation.
+
+    For example, the request to exchange a Grant Code for an Api Token.
+    """
+
+    def authenticate(self, request):
+        if not request.json_body:
+            raise AuthenticationFailed('Invalid request')
+
+        client_id = request.json_body.get('client_id')
+        client_secret = request.json_body.get('client_secret')
+
+        if not client_id or not client_secret:
```

Comment:
Going to leave this as a falsey check for now, but I agree with you @mitsuhiko; this would be better validated via json-schema (down the line 😉)

Reply:
For what it's worth there is (ugly) precedent for using jsonschema in endpoints: https://github.com/getsentry/sentry/blob/74e812774e02bd7a6f07b45f88f696a79e828405/src/sentry/api/endpoints/dif_files.py#L34-L61